man/arch/types.5
.\" @(#)types.5 1.1 94/10/31 SMI; from UCB 4.2
.TH TYPES 5 "19 October 1987"
.SH NAME
types \- primitive system data types
.SH SYNOPSIS
.B #include <sys/types.h>
.SH DESCRIPTION
.IX  "types file"  ""  "\fLtypes\fP \(em primitive system data types"
.IX  "primitive system data types"  ""  "primitive system data types \(em \fLtypes\fP"
.IX  "system data types"  ""  "system data types \(em \fLtypes\fP"
.IX  "data types"  ""  "data types \(em \fLtypes\fP"
.LP
The data types defined in the include file
are used in the
system code; some data of these types
are accessible to user code:
.LP
.nf
.ft B
.ta \w'typedef\ \ 'u +\w'minor(x)\ \ 'u +\w' short\ \ 'u

/*
 * Copyright (c) 1982, 1986 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef	_TYPES_
#define	_TYPES_

/*
 * Basic system types.
 */

#include <sys/sysmacros.h>

typedef	unsigned char	u_char;
typedef	unsigned short	u_short;
typedef	unsigned int	u_int;
typedef	unsigned long	u_long;
typedef	unsigned short	ushort;		/* System V compatibility */
typedef	unsigned int	uint;		/* System V compatibility */

#ifdef vax
typedef	struct	_physadr { int r[1]; } *physadr;
typedef	struct	label_t	{
	int	val[14];
} label_t;
#endif
#ifdef mc68000
typedef	struct	_physadr { short r[1]; } *physadr;
typedef	struct	label_t	{
	int	val[13];
} label_t;
#endif
#ifdef sparc
typedef struct  _physadr { int r[1]; } *physadr;
typedef struct label_t {
	int     val[2];
} label_t;
#endif
#ifdef i386
typedef struct	_physadr { short r[1]; } *physadr;
typedef struct	label_t {
	int	val[8];
} label_t;
#endif
typedef	struct	_quad { long val[2]; } quad;
typedef	long	daddr_t;
typedef	char *	caddr_t;
typedef	u_long	ino_t;
typedef	long	swblk_t;
typedef	int	size_t;
typedef	long	time_t;
typedef	short	dev_t;
typedef	long	off_t;
typedef	u_short	uid_t;
typedef	u_short	gid_t;
typedef	long	key_t;

#define	NBBY	8		/* number of bits in a byte */
/*
 * Select uses bit masks of file descriptors in longs.
 * These macros manipulate such bit fields (the filesystem macros use chars).
 * FD_SETSIZE may be defined by the user, but the default here
 * should be >= NOFILE (param.h).
 */
#ifndef	FD_SETSIZE
#define	FD_SETSIZE	256
#endif

typedef	long	fd_mask;
#define	NFDBITS	(sizeof(fd_mask) * NBBY)	/* bits per mask */
#ifndef	howmany
#ifdef sun386
#define howmany(x, y)   ((((u_int)(x))+(((u_int)(y))-1))/((u_int)(y)))
#else
#define	howmany(x, y)	(((x)+((y)-1))/(y))
#endif
#endif

typedef	struct fd_set {
	fd_mask	fds_bits[howmany(FD_SETSIZE, NFDBITS)];
} fd_set;

typedef	char *	addr_t;

#define	FD_SET(n, p)	((p)->fds_bits[(n)/NFDBITS] |= (1 << ((n) % NFDBITS)))
#define	FD_CLR(n, p)	((p)->fds_bits[(n)/NFDBITS] &= ~(1 << ((n) % NFDBITS)))
#define	FD_ISSET(n, p)	((p)->fds_bits[(n)/NFDBITS] & (1 << ((n) % NFDBITS)))
#define	FD_ZERO(p)	bzero((char *)(p), sizeof(*(p)))

#ifdef sparc
/*
 * routines that call setjmp have strange control flow graphs,
 * since a call to a routine that calls resume/longjmp will eventually
 * return at the setjmp site, not the original call site.  This
 * utterly wrecks control flow analysis.
 */
extern int setjmp();
#pragma unknown_control_flow(setjmp)
#endif sparc

#endif	_TYPES_
.ft R
.fi
.LP
The form
.I daddr_t
is used for disk addresses,
.\"except in an inode on disk,
see
.BR fs (5).
Times are encoded in seconds since 00:00:00
.SM GMT\s0,
January 1, 1970.
The major and minor parts of a device code
specify kind and unit number of a device
and are installation-dependent.
Offsets are measured in bytes from the
beginning of a file.  The
.I label_t
variables are used to save the processor state
while another process is running.
.SH SEE ALSO
.BR adb (1),
.BR lseek (2),
.BR time (3C),
.BR fs (5)


lang/dbx/dbx/common/types.c
#ifndef lint
static	char sccsid[] = "@(#)types.c 1.1 94/10/31 Copyr 1984 Sun Micro";
#endif

/*
 * Copyright (c) 1987 by Sun Microsystems, Inc.
 */

/*
 * Routines to manipulate the types of variables
 */

#include "defs.h"
#include "symbols.h"
#include "tree.h"
#include "names.h"

public Symbol combitypes(t1, t2)
    Symbol t1, t2;
{
    char newname[1024];
    Name n;
    Symbol s;

    strcat(strcat(strcpy(newname, idnt(t1->name)) , " "), idnt(t2->name));
    if ((n = identname(newname, false)) == nil || (s = lookup(n)) == nil)
	error("no such type");
    return s;
}

private Symbol lookup_tag(cl, tag)
    register Symclass  cl;
    Name tag;
{

    register Symbol s;

    find(s, tag) where
	s->class == TYPEID and s->type->class == cl
    endfind(s);
    if (s==nil)
	error("no such type");
    return s;
}

public Symbol lookup_structtag(tag)
    Name tag;
{
    return lookup_tag(RECORD, tag);
}

public Symbol lookup_uniontag(tag)
    Name tag;
{
    return lookup_tag(UNION, tag);
}

public Symbol lookup_enumtag(tag)
    Name tag;
{
    return lookup_tag(SCAL, tag);
}


/*
* build-abstract-declarator
*/
private Node build_abdcl(n, class)
    Node n;
    Symclass class;
{
    register Symbol *s;

    s = &n->value.sym;
    while ((*s)->class != TYPE)
	s = & (*s)->type;
    *s = newSymbol(nil, 0, class, *s, nil);
    return n;
}

public Node build_ptrdcl(n)
    Node n;
{
    return build_abdcl(n, PTR);
}

public Node build_funcdcl(n)
    Node n;
{
    return build_abdcl(n, FUNC);
}

public Node buildtype(typespec, abdcl)
    Node typespec;
    Node abdcl;
{
    /*
     * We want to make a full type out of a typespec and an
     * abstract declarator.  At the bottom of the declarator is
     * an int type.  It is only a placeholder for the real type
     * from the typespec that we want to plug in.  Do it,
     * and give back the spare pieces.
     */
    register Symbol *s, sp;
    Language clang = findlanguage(".c");
    Language modula2 = findlanguage(".mod");

    if (typespec->value.sym == t_unsigned) {
	typespec = build(O_SYM, lookup(identname("unsigned int", true)));
    }
    s = & abdcl->value.sym;
    while ((*s)->class != TYPE)
	s = & ((*s)->type);
    /* give_back_symbol(*s); */
    *s = typespec->value.sym;
    tfree(typespec);

    sp = abdcl->value.sym ;
    abdcl->nodetype = sp ;

    if( sp->class == IMPORT ) {
	    sp->language = modula2;
    } else {
	while( sp  and  sp->class != TYPE  and  sp->class != TYPEID ) {
	    sp->language = clang;
	    sp = sp->type;
	}
    }
    return abdcl;
}

/*
 * Print leading part of type
 */
public void pre_type(s)
    Symbol s;
{
    Symbol t;

    switch (s->class){
    case TYPEID:
	/* struct, union, enum */
	printf("%s %s ", c_classname(s->type),  idnt(s->name));
	break;
    case TYPE:
	printf("%s ", idnt(s->name));
	break;
    case PTR:
	if (s->type->class != BADUSE) {
	    pre_type(s->type);
	}
	if (s->type->class == FUNC or s->type->class == ARRAY) {
	    printf("(");
	}
	printf("*");
	break;
    case FUNC:
    case ARRAY:
	pre_type(s->type);
	break;
    case RECORD:
    case UNION:
    case SCAL:
	printf("%s ", c_classname(s));
	break;
    case FORWARD:
	t = lookup_forward(s);
	if (t != nil) {
	    pre_type(t);
	} else {
	    printf("%s %s ", c_tagtype(s->symvalue.forwardref.class),
	      idnt(s->symvalue.forwardref.name));
	}
	break;
    default:
	error("unexpected class cast");
    }
}

/*
 * Print trailing part of type
 */
public void post_type(s)
    Symbol s;
{
    Symbol b;

    switch (s->class){
    case TYPEID:
    case TYPE:
    case RECORD:
    case UNION:
    case SCAL:
	break;
    case PTR:
	if (s->type->class == FUNC or s->type->class == ARRAY) {
	    printf(")");
	}
	if (s->type->class != BADUSE) {
	    post_type(s->type);
	}
	break;
    case FUNC:
	printf("()");
	post_type(s->type);
	break;
    case ARRAY:
	b = rtype(s->chain);
	if (b->symvalue.rangev.upper >= 0)
	    printf("[%d]", b->symvalue.rangev.upper+1);
	else
	    printf("[]");
	post_type(s->type);
	break;
    case FORWARD:
	s = lookup_forward(s);
	if (s != nil) {
	    post_type(s);
	}
	break;
    default:
	error("unexpected class cast");
    }
}

/*
 * Print a type rename (cast) in C-like syntax
 */
public void prcast(typer)
    Node typer;
{
    printf("(");
    pre_type(typer->value.sym);
    post_type(typer->value.sym);
    printf(") ");
}

/*
 * Print out the storage class of something.
 * Either "static", "register" or nothing.
 */
public pr_storage(s)
    Symbol s;
{
    if ((char)s->level == (char)2) {
        printf("static ");
    } else if ((char)s->level < (char)0) {
        printf("register ");
    }
}
